File: backend/app/models/survey_model.py
# 📂 app/models/survey_model.py

from sqlalchemy import Column, String, ForeignKey, Text, Boolean, TIMESTAMP, Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import uuid

from app.database import Base


class Survey(Base):
    __tablename__ = "surveys"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    title = Column(String(255), nullable=False)
    description = Column(Text)
    is_active = Column(Boolean, default=True)
    created_at = Column(TIMESTAMP, server_default=func.now())

    questions = relationship("SurveyQuestion", back_populates="survey", cascade="all, delete")


class SurveyQuestion(Base):
    __tablename__ = "survey_questions"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    survey_id = Column(UUID(as_uuid=True), ForeignKey("surveys.id", ondelete="CASCADE"))
    question_text = Column(Text, nullable=False)

    survey = relationship("Survey", back_populates="questions")
    answers = relationship("SurveyAnswer", back_populates="question", cascade="all, delete")


class SurveyAnswer(Base):
    __tablename__ = "survey_answers"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    question_id = Column(UUID(as_uuid=True), ForeignKey("survey_questions.id", ondelete="CASCADE"))
    answer_text = Column(Text, nullable=False)
    votes = Column(Integer, default=0)

    question = relationship("SurveyQuestion", back_populates="answers")


class SurveyResponse(Base):
    __tablename__ = "survey_responses"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    survey_id = Column(UUID(as_uuid=True), ForeignKey("surveys.id", ondelete="CASCADE"))
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"))
    survey = relationship('Survey')
    submitted_at = Column(TIMESTAMP, server_default=func.now())


class SurveyResponseAnswer(Base):
    __tablename__ = "survey_response_answers"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    response_id = Column(UUID(as_uuid=True), ForeignKey("survey_responses.id", ondelete="CASCADE"))
    question_id = Column(UUID(as_uuid=True), ForeignKey("survey_questions.id", ondelete="CASCADE"))
    answer_id = Column(UUID(as_uuid=True), ForeignKey("survey_answers.id", ondelete="CASCADE"))

File: lib/models/survey_model.dart
class Survey {
  final String id;
  final String title;
  final String description;
  final bool isActive; // ← добавлено!
  final List<SurveyQuestion> questions;

  Survey({
    required this.id,
    required this.title,
    required this.description,
    required this.isActive,
    required this.questions,
  });

  factory Survey.fromJson(Map<String, dynamic> json) {
    return Survey(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      isActive: json['is_active'] ?? true, // ← добавлено!
      questions: (json['questions'] as List<dynamic>? ?? [])
          .map((q) => SurveyQuestion.fromJson(q))
          .toList(),
    );
  }
}

class SurveyQuestion {
  final String id;
  final String questionText;
  final List<SurveyAnswer> answers;

  SurveyQuestion({
    required this.id,
    required this.questionText,
    required this.answers,
  });

  factory SurveyQuestion.fromJson(Map<String, dynamic> json) {
    return SurveyQuestion(
      id: json['id'] ?? '',
      questionText: json['question_text'] ?? '',
      answers: (json['answers'] as List<dynamic>? ?? [])
          .map((a) => SurveyAnswer.fromJson(a))
          .toList(),
    );
  }
}

class SurveyAnswer {
  final String id;
  final String answerText;
  int votes;

  SurveyAnswer({
    required this.id,
    required this.answerText,
    this.votes = 0,
  });

  factory SurveyAnswer.fromJson(Map<String, dynamic> json) {
    return SurveyAnswer(
      id: json['id'] ?? '',
      answerText: json['answer_text'] ?? '',
      votes: json['votes'] ?? 0,
    );
  }
}

class SurveyResults {
  final String surveyTitle;
  final List<QuestionResults> results;

  SurveyResults({
    required this.surveyTitle,
    required this.results,
  });

  factory SurveyResults.fromJson(Map<String, dynamic> json) {
    return SurveyResults(
      surveyTitle: json['survey_title'] ?? '',
      results: (json['results'] as List<dynamic>? ?? [])
          .map((q) => QuestionResults.fromJson(q))
          .toList(),
    );
  }
}

class QuestionResults {
  final String question;
  final List<AnswerResult> answers;

  QuestionResults({
    required this.question,
    required this.answers,
  });

  factory QuestionResults.fromJson(Map<String, dynamic> json) {
    return QuestionResults(
      question: json['question'] ?? '',
      answers: (json['answers'] as List<dynamic>? ?? [])
          .map((a) => AnswerResult.fromJson(a))
          .toList(),
    );
  }
}

class AnswerResult {
  final String answer;
  final int count;

  AnswerResult({
    required this.answer,
    required this.count,
  });

  factory AnswerResult.fromJson(Map<String, dynamic> json) {
    return AnswerResult(
      answer: json['answer'] ?? '',
      count: json['count'] ?? 0,
    );
  }
}
